test/PackageVersionsTest/FallbackVersionsTest.php

```diff
@@ -23,37 +22,45 @@ final class FallbackVersionsTest extends TestCase
 {
     public function testWillFailWithoutValidComposerLockLocation() : void
     {
-        rename(__DIR__ . '/../../composer.lock', __DIR__ . '/../../composer.lock.backup');
+        $this->backupFile(__DIR__ . '/../../vendor/composer/installed.json');
+        $this->backupFile(__DIR__ . '/../../composer.lock');
 
-        try {
-            FallbackVersions::getVersion('phpunit/phpunit');
+        $this->expectException(UnexpectedValueException::class);
+        $this->expectExceptionMessageRegExp(
+            '@PackageVersions could not locate the `vendor/composer/installed\.json` or your `composer\.lock` '
+            . 'location\. This is assumed to be in \[[^]]+?\]\. If you customized your composer vendor directory and ran composer '
+            . 'installation with --no-scripts or if you deployed without the required composer files, then you are on '
+            . 'your own, and we can\'t really help you\. Fix your shit and cut the tooling some slack\.@'
+        );
 
-            self::fail('An exception was supposed to be thrown');
-        } catch (UnexpectedValueException $lockFileNotFound) {
-            $srcDir = realpath(__DIR__ . '/../../src/PackageVersions');
+        FallbackVersions::getVersion('phpunit/phpunit');
```

Comment:
In this scenario, you are checking for `phpunit/phpunit` against an `installed.lock` generated by `composer install --dev`. What happens if you `composer install --no-dev`? `composer.lock` will have this package, but `installed.json` won't.

Reply:
To be fair, what would've happened before this PR? I would imagine, the same thing as before since this has previously relied on the `composer.lock`. 

Do we go the extra mile and check that the package directory exists in the vendor folder? I'm okay with that if you are.

Reply:
Actually, thinking about your other comments on `FallbackVersions.php`, if we're going to separate the logic for testing the `installed.json` and `composer.lock`, we can do both checks for the package in both configurations. We first check `installed.json`, and if found we stop and continue going as expected. Otherwise, we fallback (pun intended) to the `composer.lock` configuration. If still not found, then we complain through an exception. That should satisfy `--no-dev`. Thoughts?

Reply:
I agree, indeed this already fixes most problems

Reply:
FYI, I haven't forgotten about this; I just haven't had a chance yet to implement the requested suggestions.

Reply:
That's fine: just poke me whenever you want me to go through it again, and we'll get a release sorted out.

Please also do a rebase while/when at it.

Reply:
So the logic now is to merge the contents of both the `installed.json` and the `composer.lock` so that if `--no-dev` is passed, we can still validate the package in the event that `installed.json` ends up being an empty array.